### PR TITLE
.mid: Rewrite SysEx event parsing/writing to properly use the specs defined by Phase Shift

### PR DESCRIPTION
**Pull request by TheNathannator**

### CODE REVIEW

**Comment by FireFox2000000** [comment]
Does RockBand uses SysEx events at all? Could seem a bit confusing to be labelling all this as PS when it's handling other game types as well.

**Comment by TheNathannator** [comment]
Rock Band doesn't use any SysEx events. The Phase Shift team created the ones that charts use, hence the naming.

File: Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/MidReader.cs [hunk 1]
```diff
@@ -416,142 +438,61 @@ private static void ReadNotes(IList<MidiEvent> track, Song song, Song.Instrument
...
-                                    break;
+                                processParams.midiEvent = startEvent;
+                                EventProcessFn processFn;
+                                if (processParams.sysexProcessMap.TryGetValue(psEvent.Code, out processFn))
+                                {
+                                    processFn(processParams);
                                 }
+                                break;
                             }
                         }
```
Comment (FireFox2000000) [inline Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/MidReader.cs:480]: Neater to move this search into it's own function
`static bool TryLocateStartEvent(IList<PhaseShiftSysExStart> sysexEventQueue, PhaseShiftSysEx psEndEvent, out int startEventIndex)`
And then remove from the list perform the process function if that returns true, which should be overall more readable and avoids the RemoveAt in the middle of the loop relying on the break statement after all the processing, which overall is a bit icky.

**Review by FireFox2000000 — CHANGES_REQUESTED** [review]

**Comment by FireFox2000000** [comment]
> Rock Band doesn't use any SysEx events. The Phase Shift team created the ones that charts use, hence the naming.

Cool, just wanted to confirm that.

File: Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/PhaseShiftSysEx.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+using NAudio.Midi;
+
+namespace MoonscraperChartEditor.Song.IO
+{
+    public class PhaseShiftSysEx : MidiEvent
+    {
+        public byte Type;
+        public byte Difficulty;
+        public byte Code;
+        public byte Value;
```
[the remark this answers is not in the document]
Comment (FireFox2000000) [reply Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/PhaseShiftSysEx.cs:14]: Yeah, best not to mix external library coding standards in the main codebase so I'd still want this changed. If it was part of it's own package/library I wouldn't mind.

[the remark this answers is not in the document]
Comment (TheNathannator) [reply Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/PhaseShiftSysEx.cs:14]: My problem with that is if I do that then there's still going to be mixed standards from the inherited properties, though the only one that really gets used is AbsoluteTime.
image

**Review by FireFox2000000 — APPROVED** [review]

**Comment by FireFox2000000** [comment]
Needs merge conflicts fixed.

**Comment by TheNathannator** [comment]
Figured there would be some, got those fixed up.

**Comment by TheNathannator** [comment]
And this is exactly why you test changes.
image
What I had before should really trigger a warning lol
image

**Comment by TheNathannator** [comment]
Alright, looks like it's working correctly again. That should be all, then.

**Pull request merged**